zrml/prediction-markets/src/lib.rs

```diff
@@ -1349,6 +1350,338 @@ mod pallet {
+        /// in `MarketIdsPerClose...` either at the old period end or new period end.
+        #[pallet::call_index(19)]
+        #[pallet::weight(T::WeightInfo::dispute_early_close(CacheSize::get(), CacheSize::get()))]
+        #[transactional]
+        pub fn dispute_early_close(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+            Self::ensure_market_is_active(&market)?;
+            let mut early_close = market.early_close.ok_or(Error::<T>::NoEarlyCloseScheduled)?;
+            match early_close.state {
+                EarlyCloseState::ScheduledAsMarketCreator => (),
+                EarlyCloseState::ScheduledAsOther
+                | EarlyCloseState::Disputed
+                | EarlyCloseState::Rejected => {
+                    return Err(Error::<T>::InvalidEarlyCloseState.into());
+                }
+            };
+
+            // ensure we don't dispute if the old market period is already over
+            // so that we don't switch back to an invalid market period
+            // this should never trigger, because at the time of scheduling a new market period,
+            // we ensure the new end is always before the old end
+            debug_assert!(
+                {
+                    match early_close.old {
+                        MarketPeriod::Block(ref range) => {
+                            let now_block = <frame_system::Pallet<T>>::block_number();
+                            now_block < range.end
+                        }
+                        MarketPeriod::Timestamp(ref range) => {
+                            let now_time = <zrml_market_commons::Pallet<T>>::now();
+                            now_time < range.end
+                        }
+                    }
+                },
+                "This will never happen, because schedule_early_close ensures that the new end is \
+                 always before the old end, otherwise the switch to an old market period would \
+                 lead to a never ending market!"
+            );
+
+            let close_dispute_bond = T::CloseEarlyDisputeBond::get();
+
+            T::AssetManager::reserve_named(
+                &Self::reserve_id(),
+                Asset::Ztg,
+                &who,
+                close_dispute_bond,
+            )?;
+
+            let ids_len_0 = Self::clear_auto_close(&market_id)?;
+
+            <zrml_market_commons::Pallet<T>>::mutate_market(&market_id, |market| {
+                market.period = early_close.old.clone();
+                market.bonds.close_dispute = Some(Bond::new(who.clone(), close_dispute_bond));
+                early_close.state = EarlyCloseState::Disputed;
+                market.early_close = Some(early_close);
```

Comment:
Can it happen that the market resolves at the original (first) close time, but the storage still contains `market.early_close = Some(early_close)`? In that case the final market storage element would represent an untruthful state.

Reply:
Please look here https://github.com/zeitgeistpm/zeitgeist/pull/1127#discussion_r1370084229

If the `early_close` field ever contains `state` `Disputed` or `Rejected`, it means that the early close was not applied (as the name of the state outlines). But generally I see your point here!